autopush/web/webpush.py

```diff
@@ -1,12 +1,329 @@
+    content_encoding = fields.String(
+        required=True,
+        load_from="content-encoding",
+        validate=OneOf(["aesgcm128"])
+    )
+    encryption = fields.String(required=True)
+    encryption_key = fields.String(
+        required=True,
+        load_from="encryption-key"
+    )
+    crypto_key = fields.String(load_from="crypto-key")
+
+    @validates("encryption")
+    def validate_encryption(self, value):
+        """Must contain a salt value"""
+        ck = CryptoKey(value)
+        salt = ck.get_label("salt")
+        if not salt or not VALID_BASE64_URL.match(salt):
+            raise InvalidRequest("Invalid salt value in Encryption header",
+                                 status_code=400,
+                                 errno=110)
+
+    @validates("crypto_key")
+    def validate_crypto_key(self, value):
+        """Must not contain a dh value"""
+        ck = CryptoKey(value)
+        dh = ck.get_label("dh")
+        if dh:
+            raise InvalidRequest(
+                "dh value in Crypto-Key header not valid for 01 or earlier "
+                "webpush-encryption",
+                status_code=400,
+                errno=110,
+            )
+
+    @validates("encryption_key")
+    def validate_encryption_key(self, value):
+        """Must contain a dh value"""
+        ck = CryptoKey(value)
+        dh = ck.get_label("dh")
+        if not dh or not VALID_BASE64_URL.match("dh"):
+            raise InvalidRequest("Invalid dh value in Encryption-Key header",
+                                 status_code=400,
+                                 errno=110)
+
+
+class WebPushCrypto04HeaderSchema(Schema):
+    """Validates WebPush Message Encryption
+
+    Uses draft-ietf-webpush-encryption-04 rules for validation.
+
+    """
+    content_encoding = fields.String(
+        required=True,
+        load_from="content-encoding",
+        validate=OneOf(["aesgcm"])
+    )
+    encryption = fields.String(required=True)
+    crypto_key = fields.String(
+        required=True,
+        load_from="crypto-key",
+    )
+
+    @validates("encryption")
+    def validate_encryption(self, value):
+        """Must contain a salt value"""
+        ck = CryptoKey(value)
+        salt = ck.get_label("salt")
+        if not salt or not VALID_BASE64_URL.match(salt):
+            raise InvalidRequest("Invalid salt value in Encryption header",
+                                 status_code=400,
+                                 errno=110)
+
+    @validates("crypto_key")
+    def validate_crypto_key(self, value):
+        """Must contain a dh value"""
+        ck = CryptoKey(value)
+        dh = ck.get_label("dh")
+        if not dh or not VALID_BASE64_URL.match("dh"):
+            raise InvalidRequest("Invalid dh value in Encryption-Key header",
+                                 status_code=400,
+                                 errno=110)
+
+    @validates_schema(pass_original=True)
+    def check_unknown_fields(self, data, original_data):
```

Comment:
ah one more thing, let's clarify this method name. 'reject_encryption_key' would work